cranelift/codegen/src/machinst/isle.rs

```diff
@@ -1206,13 +1206,19 @@ macro_rules! isle_prelude_method_helpers {
                     self.lower_ctx.emit(inst);
                 }
             }
+
             // Handle retvals prior to emitting call, so the
             // constraints are on the call instruction; but buffer the
             // instructions till after the call.
             let mut outputs = InstOutput::new();
             let mut retval_insts: crate::machinst::abi::SmallInstVec<_> = smallvec::smallvec![];
-            for i in 0..num_rets {
-                let ret = self.lower_ctx.sigs()[abi].get_ret(i);
+            // We take the *last* `num_rets` returns of the sig:
+            // this skips a StructReturn, if any, that is present.
+            let sigdata = &self.lower_ctx.sigs()[abi];
+            debug_assert!(num_rets <= sigdata.num_rets());
+            for i in (sigdata.num_rets() - num_rets)..sigdata.num_rets() {
+                let sigdata = &self.lower_ctx.sigs()[abi];
```

Comment:
Did this have to be named again to avoid holding a reference to self through the body of the loop?

Reply:
Yep, exactly, this was a borrow-checker appeasement measure. I'll add a comment to that effect.